### DISCUSSION TITLE
GPIO Pin interrupt example code

### DISCUSSION THREAD

**Discussion by scottlbaker** [earlephilhower/arduino-pico#716] [Q&A]
Hi,

I've been trying to get GPIO interrupts to work. I have tried both the Arduino style code: attachInterrupt(digitalPinToInterrupt(ROTA), readEncoder, RISING); and also the SDK call: gpio_set_irq_enabled_with_callback(ROTA, 0x0c, 1, &readEncoder); and neither of these seems to work. My ISR never gets called. Can anyone point me to some example working GPIO interrupt code. 

Thanks,
Scott

**Comment by scottlbaker** [comment]
Here is my current code. It does not work and it seems that the ISR is not getting called.

Thanks,
Scott

// ============================================================================
//  encoder ::  rotary encoder interrupt test
//
//  Arduino IDE settings:
//  board: Raspberry Pi Pico (Arduino-Pico)
//  programmer: none
//
// ============================================================================

#define TXD1       8    // GPIO8   PIN11  UART1 Tx
#define RXD1       9    // GPIO9   PIN12  UART1 Rx
#define ROTA       6    // GPIO6   PIN8   rotary encoder A
#define ROTB       7    // GPIO7   PIN9   rotary encoder B
#define DIAG      25    // GPIO25  PIN37  Diagnotic LED

#define GPIO    0x0d    // GPIO interrupt number
#define CHANGE  0x0c    // pos or neg edge triggered

int8_t  encoder_val = 0;
uint8_t encoder_state;

// rotary encoder pin change interrupt handler
void readEncoder(uint gpio, uint32_t events) {
  encoder_state = (encoder_state << 4) | (digitalRead(ROTB) << 1) | digitalRead(ROTA);
  switch (encoder_state) {
    case 0x23:  encoder_val++; break;
    case 0x32:  encoder_val--; break;
    default: break;
  }
  irq_clear(GPIO);
}

// rotary encoder init
void encoder_init() {
  gpio_set_irq_enabled_with_callback(ROTA, CHANGE, 1, &readEncoder);
  encoder_state = (digitalRead(ROTB) << 1) | digitalRead(ROTA);
  interrupts();
}

// startup initialization
void setup() {
  digitalWrite(DIAG, LOW);       // Diag LED is off
  pinMode(DIAG, OUTPUT);
  pinMode(ROTA, INPUT_PULLUP);
  pinMode(ROTB, INPUT_PULLUP);
  #define BAUD  115200
  Serial2.setRX(RXD1);
  Serial2.setTX(TXD1);
  Serial2.begin(BAUD);
  irq_set_enabled(GPIO, 1);
}

// main program loop
void loop() {
  if (encoder_val) {
    Serial2.println(encoder_val);
  }
  delay(10);
}

**Reply by earlephilhower** [reply]
I suggest using the Arduino-way and not the SDK-way.  See https://github.com/earlephilhower/arduino-pico/issues/651 where a bug was found and fixed in the Pico SDK (many releases ago) for an example

**Reply by scottlbaker** [reply]
Hi,

Thanks for the response.
I rolled it back to use the Arduino way and it still doesn't work.
The LED never turns on, so the ISR is never getting called.

Scott

// ============================================================================
//  rotary encoder interrupt test
// ============================================================================

#define TXD1       8    // GPIO8   PIN11  UART1 Tx
#define RXD1       9    // GPIO9   PIN12  UART1 Rx
#define ROTA       6    // GPIO6   PIN8   rotary encoder A
#define ROTB       7    // GPIO7   PIN9   rotary encoder B
#define DIAG      25    // GPIO25  PIN37  diagnostic LED

int8_t  encoder_val = 0;
uint8_t encoder_state;

// rotary encoder pin change interrupt handler
void readEncoder() {
  digitalWrite(DIAG, HIGH);      // diag LED is on
  encoder_state = (encoder_state << 4) | (digitalRead(ROTB) << 1) | digitalRead(ROTA);
  switch (encoder_state) {
    case 0x23:  encoder_val++; break;
    case 0x32:  encoder_val--; break;
    default: break;
  }
}

// rotary encoder init
void encoder_init() {
  attachInterrupt(digitalPinToInterrupt(ROTA), readEncoder, RISING);
  encoder_state = (digitalRead(ROTB) << 1) | digitalRead(ROTA);
}

// startup initialization
void setup() {
  digitalWrite(DIAG, LOW);       // diag LED is off
  pinMode(DIAG, OUTPUT);
  pinMode(ROTA, INPUT_PULLUP);
  pinMode(ROTB, INPUT_PULLUP);
  #define BAUD  115200
  Serial2.setRX(RXD1);
  Serial2.setTX(TXD1);
  Serial2.begin(BAUD);
  delay(100);
}

// main program loop
void loop() {
  if (encoder_val) {
    Serial2.println(encoder_val);
  }
  delay(10);
}

**Reply by scottlbaker** [reply]
Oops .. 1000 apologies .. typo on my part .. working now ..

Thanks !!
Scott

**Reply by earlephilhower** [reply]
Great, glad to hear it's working!
